src/functional/geometry/convex_hull.zig

```diff
@@ -0,0 +1,167 @@
+
+    const cross_product = (middle_point.value - first_point.value) * distance_last_middle -
+        (last_point.value - middle_point.value) * distance_middle_first;
+
+    return if (cross_product == 0) Turn.collinear else if (cross_product > 0) Turn.right else Turn.left;
+}
+
+test "incremental convex hull with known result" {
+    const allocator = testing.allocator;
+
+    var convex_hull = try ConvexHull.init(allocator);
+    defer convex_hull.deinit();
+
+    try convex_hull.addPoint(.{ .time = 0, .value = 3 });
+    try convex_hull.addPoint(.{ .time = 1, .value = 2 });
+    try convex_hull.addPoint(.{ .time = 2, .value = 3.5 });
+    try convex_hull.addPoint(.{ .time = 3, .value = 5 });
+    try convex_hull.addPoint(.{ .time = 4, .value = 3 });
+    try convex_hull.addPoint(.{ .time = 5, .value = 4 });
+    try convex_hull.addPoint(.{ .time = 6, .value = 4 });
+    try convex_hull.addPoint(.{ .time = 7, .value = 3 });
+    try convex_hull.addPoint(.{ .time = 8, .value = 4.5 });
+    try convex_hull.addPoint(.{ .time = 9, .value = 3.5 });
+    try convex_hull.addPoint(.{ .time = 10, .value = 2.5 });
+    try convex_hull.addPoint(.{ .time = 11, .value = 2.5 });
+    try convex_hull.addPoint(.{ .time = 12, .value = 3.5 });
+    try convex_hull.addPoint(.{ .time = 13, .value = 2.5 });
+    try convex_hull.addPoint(.{ .time = 14, .value = 2.5 });
+    try convex_hull.addPoint(.{ .time = 15, .value = 2.5 });
+    try convex_hull.addPoint(.{ .time = 16, .value = 3 });
+    try convex_hull.addPoint(.{ .time = 17, .value = 3 });
+    try convex_hull.addPoint(.{ .time = 18, .value = 3 });
+    try convex_hull.addPoint(.{ .time = 19, .value = 3 });
+    try convex_hull.addPoint(.{ .time = 20, .value = 2.8 });
+
+    try testing.expectEqual(5, convex_hull.upper_hull.items.len);
+    try testing.expectEqual(4, convex_hull.lower_hull.items.len);
+
+    // Expected Upper Hull.
+    try testing.expectEqual(0, convex_hull.upper_hull.items[0].time);
+    try testing.expectEqual(3, convex_hull.upper_hull.items[1].time);
+    try testing.expectEqual(8, convex_hull.upper_hull.items[2].time);
+    try testing.expectEqual(19, convex_hull.upper_hull.items[3].time);
+    try testing.expectEqual(20, convex_hull.upper_hull.items[4].time);
+
+    // Expected Lower Hull.
+    try testing.expectEqual(0, convex_hull.lower_hull.items[0].time);
+    try testing.expectEqual(1, convex_hull.lower_hull.items[1].time);
+    try testing.expectEqual(15, convex_hull.lower_hull.items[2].time);
+    try testing.expectEqual(20, convex_hull.lower_hull.items[3].time);
+}
+
+test "incremental convex hull random elements" {
+    const num_points: usize = 1000;
+    const allocator = testing.allocator;
+    var rnd = std.rand.DefaultPrng.init(0);
+
+    var convex_hull = try ConvexHull.init(allocator);
+    defer convex_hull.deinit();
+
+    for (0..num_points) |i|
+        try convex_hull.addPoint(.{ .time = i, .value = rnd.random().float(f64) });
```

Comment:
Maybe include {} around the loop body to minimize mistakes if additional lines of code are added.

Reply:
Fixed.